File: scripts/csharp_controller/Options.cs
// Options.cs - Command-line options, same flags as controller.py / the C++ controller.

using System.Globalization;

namespace ArmController;

public enum ControlMode { Joystick, Slider, Ik, Fleet }

public sealed class Options
{
    public static Options Current { get; set; } = new();

    public int Device { get; set; }
    public string? Port { get; set; }
    public int Baud { get; set; } = 115200;
    public double Rate { get; set; } = 30;
    public ControlMode? Mode { get; set; }
    public bool ListJoysticks { get; set; }
    public bool ListPorts { get; set; }
    public int FleetPort { get; set; } = FleetServer.DefaultPort;
    public string RiftHost { get; set; } = "127.0.0.1";
    public int RiftPort { get; set; } = 5000;
    public bool NoRegister { get; set; }
    public bool Serve { get; set; }
    public string? Connect { get; set; }

    public static bool TryParse(string[] args, out Options options, out string error)
    {
        options = new Options();
        string? problem = Parse(args, options);
        error = problem ?? "";
        return problem == null;
    }

    static string? Parse(string[] args, Options o)
    {
        for (int i = 0; i < args.Length; i++)
        {
            string a = args[i];

            string? value = null;
            if (a is "--port" or "--connect" or "--rift-host" or "--device" or "--baud" or "--fleet-port" or "--rift-port" or "--rate" or "--mode")
            {
                if (i + 1 >= args.Length) return $"{a} requires a value";
                value = args[++i];
            }

            switch (a)
            {
                case "--list": o.ListJoysticks = true; break;
                case "--list-ports": o.ListPorts = true; break;
                case "--no-register": o.NoRegister = true; break;
                case "--serve": o.Serve = true; break;
                case "--connect": o.Connect = value; break;
                case "--port": o.Port = value; break;
                case "--rift-host": o.RiftHost = value!; break;
                case "--device":
                    if (!int.TryParse(value, out int device)) return $"{a} requires an integer";
                    o.Device = device;
                    break;
                case "--baud":
                    if (!int.TryParse(value, out int baud)) return $"{a} requires an integer";
                    o.Baud = baud;
                    break;
                case "--fleet-port":
                    if (!int.TryParse(value, out int fleetPort)) return $"{a} requires an integer";
                    o.FleetPort = fleetPort;
                    break;
                case "--rift-port":
                    if (!int.TryParse(value, out int riftPort)) return $"{a} requires an integer";
                    o.RiftPort = riftPort;
                    break;
                case "--rate":
                    if (!double.TryParse(value, NumberStyles.Float, CultureInfo.InvariantCulture, out double rate))
                        return $"{a} requires a number";
                    o.Rate = rate;
                    break;
                case "--mode":
                    switch (value)
                    {
                        case "joystick": o.Mode = ControlMode.Joystick; break;
                        case "slider": o.Mode = ControlMode.Slider; break;
                        case "ik": o.Mode = ControlMode.Ik; break;
                        case "fleet": o.Mode = ControlMode.Fleet; break;
                        default: return $"Unknown --mode '{value}' (expected joystick, slider, ik or fleet)";
                    }
                    break;
                default:
                    return $"Unknown argument: {a}";
            }
        }
        if (o.Connect != null && (o.Serve || o.Mode == ControlMode.Fleet))
            return "--connect can't be combined with --serve or --mode fleet (this controller has no arm of its own)";
        return null;
    }
}

File: scripts/rust_controller/src/options.rs
//! Command-line options, same flags as controller.py / the other controllers.

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Mode {
    Joystick,
    Slider,
    Ik,
    Fleet,
}

#[derive(Clone, Debug)]
pub struct Options {
    pub device: i32,
    pub port: Option<String>,
    pub baud: u32,
    pub rate: f64,
    pub mode: Option<Mode>,
    pub list_joysticks: bool,
    pub list_ports: bool,
    pub fleet_port: u16,
    pub rift_host: String,
    pub rift_port: u16,
    pub no_register: bool,
    pub serve: bool,
    pub connect: Option<String>,
}

impl Default for Options {
    fn default() -> Self {
        Self {
            device: 0,
            port: None,
            baud: 115200,
            rate: 30.0,
            mode: None,
            list_joysticks: false,
            list_ports: false,
            fleet_port: crate::fleet::DEFAULT_PORT,
            rift_host: "127.0.0.1".into(),
            rift_port: 5000,
            no_register: false,
            serve: false,
            connect: None,
        }
    }
}

impl Options {
    pub fn parse(args: impl IntoIterator<Item = String>) -> Result<Self, String> {
        let mut o = Options::default();
        let mut it = args.into_iter();
        while let Some(a) = it.next() {
            let mut value = |flag: &str| it.next().ok_or_else(|| format!("{flag} requires a value"));
            match a.as_str() {
                "--list" => o.list_joysticks = true,
                "--list-ports" => o.list_ports = true,
                "--no-register" => o.no_register = true,
                "--serve" => o.serve = true,
                "--port" => o.port = Some(value(&a)?),
                "--connect" => o.connect = Some(value(&a)?),
                "--rift-host" => o.rift_host = value(&a)?,
                "--device" => o.device = num(&a, &value(&a)?)?,
                "--baud" => o.baud = num(&a, &value(&a)?)?,
                "--fleet-port" => o.fleet_port = num(&a, &value(&a)?)?,
                "--rift-port" => o.rift_port = num(&a, &value(&a)?)?,
                "--rate" => o.rate = num(&a, &value(&a)?)?,
                "--mode" => {
                    o.mode = Some(match value(&a)?.as_str() {
                        "joystick" => Mode::Joystick,
                        "slider" => Mode::Slider,
                        "ik" => Mode::Ik,
                        "fleet" => Mode::Fleet,
                        other => return Err(format!("Unknown --mode '{other}' (expected joystick, slider, ik or fleet)")),
                    })
                }
                other => return Err(format!("Unknown argument: {other}")),
            }
        }
        if o.connect.is_some() && (o.serve || o.mode == Some(Mode::Fleet)) {
            return Err("--connect can't be combined with --serve or --mode fleet (this controller has no arm of its own)".into());
        }
        Ok(o)
    }
}

fn num<T: std::str::FromStr>(flag: &str, text: &str) -> Result<T, String> {
    text.parse().map_err(|_| format!("{flag} requires a number"))
}
